### PR TITLE
[CORE] Change get_input_const_data_as return type from std::unique_ptr to ov::optional

### PR DESCRIPTION
**Pull request by barnasm1**
### Details:
 - Return type replaced
 - Checked if binary size decreased.

### Tickets:
 - CVS-111396

### CODE REVIEW

**Review by praasz — CHANGES_REQUESTED** [review]
Fix issues in tests

File: src/core/shape_inference/include/utils.hpp [hunk 1]
```diff
@@ -204,21 +204,21 @@ template <class TShape,
...
         const auto& et = constant->get_element_type();
         const auto& shape = constant->get_shape();
-        return std::unique_ptr<TRes>(new TRes(get_raw_data_as<TData, TRes>(et,
-                                                                           constant->get_data_ptr(),
-                                                                           shape_size(shape),
-                                                                           std::forward<UnaryOperation>(func))));
+        return ov::optional<TRes>(get_raw_data_as<TData, TRes>(et,
+                                                               constant->get_data_ptr(),
+                                                               shape_size(shape),
+                                                               std::forward<UnaryOperation>(func)));
```
Comment (t-jankowski) [inline src/core/shape_inference/include/utils.hpp:221]: Inconsistent style comparing to line 212 above.

File: src/core/shape_inference/include/utils.hpp [hunk 2]
```diff
@@ -245,20 +245,20 @@ template <class TShape,
...
-                                              size_t idx,
-                                              const ITensorAccessor& tensor_accessor,
-                                              UnaryOperation&& func = ov::util::Cast<TData>()) {
+ov::optional<TRes> get_input_const_data_as(const ov::Node* op,
+                                           size_t idx,
+                                           const ITensorAccessor& tensor_accessor,
+                                           UnaryOperation&& func = ov::util::Cast<TData>()) {
     if (auto t = tensor_accessor(idx)) {
-        return std::unique_ptr<TRes>(new TRes(get_tensor_data_as<TData, TRes>(t, std::forward<UnaryOperation>(func))));
+        return ov::optional<TRes>(get_tensor_data_as<TData, TRes>(t, std::forward<UnaryOperation>(func)));
```
Comment (t-jankowski) [inline src/core/shape_inference/include/utils.hpp:253]: What about style like in line 212 above.
```suggestion
        return {get_tensor_data_as<TData, TRes>(t, std::forward<UnaryOperation>(func))};
```

**Review by t-jankowski — APPROVED** [review]

File: src/core/shape_inference/include/utils.hpp (hunk 2 above)
Comment (praasz) [reply src/core/shape_inference/include/utils.hpp:253]: This is under investigation as proposed suggestion can suppress RVO and maybe return to `ov::optional<>()` would be better option.

Comment (barnasm1) [reply src/core/shape_inference/include/utils.hpp]: Analysis show that in current implementation of `ov::optional` both options take advantage of RVO. I have decided to use shorter syntax.

**Review by t-jankowski — APPROVED** [review]

**Pull request merged**